### ISSUE TITLE
NullReferenceException when switching to different wireless network

### ISSUE THREAD

**Issue by AudioW** [onovotny/Zeroconf#16]
Hi,

Zeroconf gives a null reference exception when you use it on a wireless network adapter and you switch to a different wireless network.

**Comment by onovotny** [comment]
Hi @AudioW,

Do you have a more detailed repro? Where you in the middle of a search and then switched? Was this WPF/Desktop, Windows Phone? What OS? Do you have a call stack?

**Comment by AudioW** [comment]
Hi @onovotny ,

the exception happens when you turn one network adapter on and another on off (switching between adapters) and when you switch between wireless networks. I am running this on Windows 10 Store app.

I'll try and get a call stack up here..

**Comment by AudioW** [comment]
Hi @onovotny ,

I am in the middle of a search when switching between networks (or adapters).

this is the call stack:

 	mscorlib.ni.dll!System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task task)	Unknown

I cannot get any more info out of VS for some reason. Is there any other info I can supply you with?

Thanks

**Comment by AudioW** [comment]
at Zeroconf.NetworkInterface.<BindToSocketAndWriteQuery>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Zeroconf.NetworkInterface.<NetworkRequestAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Zeroconf.ZeroconfResolver.<ResolveInternal>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Zeroconf.ZeroconfResolver.<ResolveAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()

**Comment by onovotny** [comment]
Thanks. This is with the latest version on NuGet? I'll see if I can trace down the underlying code that generated d__1 and add some more error handling. There's clearly a race condition occurring in this case.

What would you expect the behavior to be? At best, I can catch and rethrow some other exception, but I'm not sure there's much I do. Would you expect it to retry somehow or could you simply call the find method again?

**Comment by AudioW** [comment]
Hi,

yes this is with the latest version on Nuget. My guess is that the expected behavior would be for ZeroConf to reset its search and try again using the newly selected wifi network (if any) or using the network adapter that is the best option after one is turned on or off. In any case, I think it would be best if it resolves everything itself instead of the user having to bother about the switching scenarios.

For me personally, I do continuous usage of the find method to keep a list of Bonjour devices synced. It wouldn't be too difficult to simply call the method again but I think the fancy way would be for Zeroconf to adapt to a new network situation.

Thanks

**Comment by ekwus** [comment]
Hi

I would suggest that the expected behaviour would be to throw a well defined exception and stop.

To retry is doing application level functionality at the library level. What if the interface were to go away for good, how long do you retry?

It is much more sensible to make the retry decision at the app level.

Cheers

Dave


Sent from Samsung Mobile



-------- Original message --------

**Comment by onovotny** [comment]
This is a tough one as it's hard to tell where the NRE is specifically. That said, I did add an extra check that may help in this scenario. It is possible a different exception is thrown instead though. Any exceptions coming from that method though should be handled as part of the retry loop -- so my suggestion is to increase the number of retries and combine it with the retry delay to cover the time it takes to switch adapters.

The fix I did will be in 2.4.
